src/ir/auth_logic/lowering_ast_datalog.cc

```diff
@@ -262,11 +273,180 @@ std::vector<datalog::Rule> LoweringToDatalogPass::QueriesToDLIR(
-    datalog::Predicate lhs(query.name(), {"dummy_var"}, datalog::kPositive);
+    datalog::Predicate lhs(query.name(), {R"("dummy_var")"},
+                           datalog::kPositive);
     return datalog::Rule(lhs, {main_pred, kDummyPredicate});
   });
 }
 
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::TransformDeclarations(
+    const std::vector<datalog::RelationDeclaration>& relation_declarations) {
+  // Attributes have a principal that the attribute is applied to
+  // as an additional parameter at the beginning of the param list.
+  // Example below shows transforming the attributes.
+  // Ex: .decl ATTRIBUTE hasProperty(x0 : Symbol) -> .decl
+  // hasProperty(attribute_prin : Prinipal, x0 : Symbol)
+  return utils::MapIter<datalog::RelationDeclaration>(
+      relation_declarations,
+      [](const datalog::RelationDeclaration& declaration) {
+        if (declaration.is_attribute()) {
+          std::vector<datalog::Argument> transformed_arguments = {
+              datalog::Argument(
+                  "attribute_prin",
+                  datalog::ArgumentType(datalog::ArgumentType::Kind::kPrincipal,
+                                        "Principal"))};
+          absl::c_copy(declaration.arguments(),
+                       std::back_inserter(transformed_arguments));
+
+          return datalog::RelationDeclaration(declaration.relation_name(),
+                                              false, transformed_arguments);
+        } else {
+          return declaration;
+        }
+      });
+}
+
+/*
+Example: A relation declaration and canSay fact shown below.
+".decl grantAccess(x0 : symbol, x1 : symbol)
+PrincipalA cansay PrincipalB grantAccess(secretFile)."
+can_say_depth_(A hashmap) maps "grantAccess" predicate name to depth 2.
+type_environment is a mapping from the relation names to the declaration
+datalog_ir node. In this example, type_environment[grantAccess] maps to
+RelationDeclaration(grantAccess, false, ...).
+The method adds 2 new declarations
+.decl canSay_grantAccess(delegatee1 : Principal, x0 : symbol, x1 : symbol) and
+.decl canSay_canSay_grantAccess(delegatee2 : Principal, delegatee1: Principal,
+x0 : symbol, x1 : symbol)
+*/
+std::vector<datalog::RelationDeclaration>
+LoweringToDatalogPass::CanSayDeclarations(
+    const absl::flat_hash_map<std::string_view, datalog::RelationDeclaration>&
+        type_environment) {
+  // Generate relation declarations for delegated predicates with
+  // each encountered delegation depth
+  std::vector<datalog::RelationDeclaration> can_say_declarations;
+  for (const auto& [predicate_name, depth] : can_say_depth_) {
+    if (depth == 0) continue;
+    // find relation declaration corresponding to the predicate name
+    auto type_iterator = type_environment.find(predicate_name);
+    if (type_iterator != type_environment.end()) continue;
+    datalog::RelationDeclaration declaration = type_iterator->second;
```

Comment:
`const & declaration`?

Reply:
`declaration` here is being modified n the for loop below. We assign a new `RelationDeclaration` in each loop.

Reply:
Are you essentially arguments of all predicates in the `canSay` chain? Is this referring to the relation corresponding to the `canSay` chain that has been build so far? If so, how about you just hold onto the address as follows: 

```C++
const datalog::RelationDeclaration* can_say_chain_suffix = &type_iterator->second;
```

You are making an unncessary copy with `datalog::RelationDeclaration declaration = type_iterator->second;`.

Reply:
I actually modified code to remove the need for this declaration. As per these changes, no new copies or local variables are needed. Left comments to make sure these changes are easy to understand. Please let

Reply:
I actually modified code to remove the need for this declaration. As per these changes, no new copies or local variables are needed. Left comments to make sure these changes are easy to understand. Please let